src/pal/src/include/pal/sharedmemory.h

```diff
@@ -0,0 +1,247 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _PAL_SHARED_MEMORY_H_
+#define _PAL_SHARED_MEMORY_H_
+
+#include "corunix.hpp"
+
+#ifndef static_assert_no_msg
+#define static_assert_no_msg( cond ) static_assert( cond, #cond )
+#endif // !static_assert_no_msg
+
+#ifndef _countof
+#define _countof(a) (sizeof(a) / sizeof(a[0]))
+#endif // !_countof
+
+#ifndef PAGE_SIZE
+#define PAGE_SIZE (0x1000)
+#endif // !PAGE_SIZE
+
+// - Global shared memory files go in:
+//     /tmp/coreclr/shm/global/<fileName>
+// - Session-scoped shared memory files go in:
+//     /tmp/coreclr/shm/session<sessionId>/<fileName>
+// - Lock files associated with global shared memory files go in:
+//     /tmp/coreclr/lockfiles/global/<fileName>
+// - Lock files associated with session-scoped shared memory files go in:
+//     /tmp/coreclr/lockfiles/session<sessionId>/<fileName>
+
+#define SHARED_MEMORY_MAX_FILE_NAME_CHAR_COUNT (_MAX_FNAME - 1)
+#define SHARED_MEMORY_MAX_NAME_CHAR_COUNT (_countof("Global\\") - 1 + SHARED_MEMORY_MAX_FILE_NAME_CHAR_COUNT)
+
+#define SHARED_MEMORY_TEMP_DIRECTORY_PATH "/tmp"
+#define SHARED_MEMORY_RUNTIME_TEMP_DIRECTORY_PATH "/tmp/coreclr"
+
+#define SHARED_MEMORY_SHARED_MEMORY_DIRECTORY_PATH "/tmp/coreclr/shm"
+#define SHARED_MEMORY_LOCK_FILES_DIRECTORY_PATH "/tmp/coreclr/lockfiles"
```

Comment:
I looked at TMPDIR, there's also GetTempPathA in PAL that uses it. I'm not comfortable using an environment variable though. For backward compatibility (even if it means to deterministically be incompatible with a previous version after a breaking change), it needs to be a non-configurable, system-wide location. It wouldn't be possible to change the location easily without breaking back-compat.

Regarding the path, I searched for uses of /tmp in coreclr and didn't find any specifics. It's a good point though. Considering the /tmp/dotnet base used by CoreFX, I'm thinking maybe I can use /tmp/dotnet/shm and /tmp/dotnet/shm/lockfiles such that it is not specific to coreclr (perhaps corert or other runtimes will also share the sync system). Or if that's not an issue I can follow the current pattern and use /tmp/dotnet/coreclr/*. What do you think?